framework/src/restart/RestartableDataIO.C

```diff
@@ -58,63 +53,163 @@ RestartableDataIO::writeRestartableData(std::string base_file_name, const Restar
-           it != restartable_data.end();
-           ++it)
-      {
-        // Moose::out<<"Storing "<<it->first<<std::endl;
+      // Store the size of the data then the data
+      unsigned int data_size = static_cast<unsigned int>(data.tellp());
+      data_blk.write((const char *) &data_size, sizeof(data_size));
+      data_blk << data.str();
+    }
 
-        std::ostringstream data;
-        it->second->store(data);
+    // Write out this proc's block size
+    unsigned int data_blk_size = static_cast<unsigned int>(data_blk.tellp());
+    stream.write((const char *) &data_blk_size, sizeof(data_blk_size));
 
-        // Store the size of the data then the data
-        unsigned int data_size = static_cast<unsigned int>(data.tellp());
-        data_blk.write((const char *) &data_size, sizeof(data_size));
-        data_blk << data.str();
-      }
+    // Write out the values
+    stream << data_blk.str();
+  }
+}
 
-      // Write out this proc's block size
-      unsigned int data_blk_size = static_cast<unsigned int>(data_blk.tellp());
-      out.write((const char *) &data_blk_size, sizeof(data_blk_size));
+void
+RestartableDataIO::deserializeRestartableData(const std::map<std::string, RestartableDataValue *> & restartable_data, std::istream & stream, const std::set<std::string> & recoverable_data)
+{
+  bool recovering = _fe_problem.getMooseApp().isRecovering();
+
+  std::vector<std::string> ignored_data;
+
+  // number of data
+  unsigned int n_data = 0;
+  stream.read((char *) &n_data, sizeof(n_data));
+
+  // data names
+  std::vector<std::string> data_names(n_data);
+
+  for (unsigned int i=0; i < n_data; i++)
+  {
+    std::string data_name;
+    char ch = 0;
+    do {
+      stream.read(&ch, 1);
+      if (ch != '\0')
+        data_name += ch;
+    } while (ch != '\0');
+    data_names[i] = data_name;
+  }
+
+  // Grab this processor's block size
+  unsigned int data_blk_size = 0;
+  stream.read((char *) &data_blk_size, sizeof(data_blk_size));
+
+  for (unsigned int i=0; i < n_data; i++)
+  {
+    std::string current_name = data_names[i];
+
+    unsigned int data_size = 0;
+    stream.read((char *) &data_size, sizeof(data_size));
+
+    // Determine if the current data is recoverable
+    bool is_data_restartable = restartable_data.find(current_name) != restartable_data.end();
+    bool is_data_recoverable = recoverable_data.find(current_name) != recoverable_data.end();
+    if (is_data_restartable // Only restore values if they're currently being used
+        && (recovering || !is_data_recoverable)) // Only read this value if we're either recovering or this hasn't been specified to be recovery only data
+
+    {
+      // Moose::out<<"Loading "<<current_name<<std::endl;
 
-      // Write out the values
-      out << data_blk.str();
+      try
+      {
+        RestartableDataValue * current_data = restartable_data.at(current_name);
+        current_data->load(stream);
+      }
+      catch(...)
```

Comment:
Just curious did you hit this catch case? Is it just a normal seek past end of file error? If so we could probably just catch stream errors instead of catch all.